pkg/termcorder/termcorder.go

```diff
@@ -0,0 +1,210 @@
+	QuietMode     bool
+	Append        bool
+	Interactive   bool
+	LogKeystrokes bool
+	PrintHelp     bool
+}
+
+const defaultFileName = "termcording"
+
+var cli *flag.FlagSet
+
+//Cmd sets the termcording's cmd.
+func Cmd(c *exec.Cmd) func(*Termcording) error {
+	return func(tc *Termcording) error {
+		tc.cmd = c
+		return nil
+	}
+}
+
+//Output sets the termcording's output.
+func Output(w io.Writer) func(*Termcording) error {
+	return func(tc *Termcording) error {
+		tc.out = w
+		return nil
+	}
+}
+
+//NewTermcording returns a pointer to a new variable of type Termcording given a config
+//variable and (functional) options.
+func NewTermcording(c *Config, options ...func(*Termcording) error) (*Termcording, error) {
+	tc := &Termcording{
+		Config: c,
+	}
+	for _, option := range options {
+		err := option(tc)
+		if err != nil {
+			return &Termcording{}, err
+		}
+	}
+	return tc, nil
+}
+
+//TermcordingFromFlags parses command line flags (and arguments) and returns a pointer to a
+//new variable of type `Termcording`.
+func TermcordingFromFlags(options ...func(*Termcording) error) (*Termcording, error) {
+	var fName, cmdName string
+	var cmdArgs []string
+
+	var h, q, a, k bool
+
+	cli = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	cli.BoolVar(&h, "h", false, "Prints this message")
+	cli.BoolVar(&q, "q", false, "Quiet mode - suppresses the recording start and end prompts")
+	cli.BoolVar(&a, "a", false, "Appends to file instead of overwriting it")
+	cli.BoolVar(&k, "k", false, "Log key strokes to file as well")
+
+	cli.Parse(os.Args[1:])
+
+	shell, ok := os.LookupEnv("SHELL")
+	if cli.Arg(1) == "" && (!ok || shell == "") {
+		return &Termcording{}, errors.New("shell not set")
+	}
+	interactive := true
+	switch cli.NArg() {
+	case 0:
+		fName = defaultFileName
+		cmdName = shell
+	case 1:
+		fName = cli.Arg(0)
+		cmdName = shell
+	default:
+		fName = cli.Arg(0)
+		cmdName = cli.Arg(1)
+		cmdArgs = cli.Args()[2:]
+		interactive = false
+	}
+
+	return NewTermcording(&Config{
+		Filename:      fName,
+		CmdName:       cmdName,
+		CmdArgs:       cmdArgs,
+		QuietMode:     q,
+		Append:        a,
+		Interactive:   interactive,
+		LogKeystrokes: k,
+		PrintHelp:     h,
+	}, options...)
+}
+
+//Start runs the tc.cmd in a pseudo-terminal and writes all output to tc.out
+func (tc *Termcording) Start() error {
+	if tc.Config.PrintHelp {
+		if tc.out == nil {
+			tc.out = os.Stdout
+		}
+		printHelp(tc)
+		return nil
+	}
+
+	if tc.cmd == nil {
+		tc.cmd = exec.Command(tc.Config.CmdName, tc.Config.CmdArgs...)
+	}
+
+	if !tc.Config.QuietMode {
+		fmt.Println("Starting recording session. CTRL-D to end.")
+		defer fmt.Printf("\nRecording session ended. Session saved to %s\n", tc.Config.Filename)
+	}
+
+	ptmx, restoreMode, err := ptmxFromCmd(tc.cmd, tc.Config.Interactive) //TODO convert to method.
+	if err != nil {
+		return err
+	}
+	defer func() {
+		restoreMode()
+		ptmx.Close()
+	}()
+
+	in := io.Writer(ptmx)
+	if tc.out == nil {
+		mode := os.O_TRUNC
```

Comment:
Whenever we have a long `if` block it's a good sign that the code inside it is a bit too low-level. In order to keep the abstraction level of `Start` fairly high, we might think about pushing the 'setup output' bit into its own helper function/method.